libmisc/random.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * SPDX-FileCopyrightText:  Alejandro Colomar <alx@kernel.org>
+ *
+ * SPDX-License-Identifier:  BSD-3-Clause
+ */
+
+#include <config.h>
+
+#ident "$Id$"
+
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#if HAVE_SYS_RANDOM_H
+#include <sys/random.h>
+#endif
+#include "bit.h"
+#include "prototypes.h"
+#include "shadowlog.h"
+
+
+/*
+ * Return a uniformly-distributed random u_long value.
+ *
+ * Similar to arc4random(3), but returns a u_long.
+ */
+unsigned long
+shadow_random(void)
+{
+	FILE           *fp;
+	unsigned long  r;
+
+#ifdef HAVE_GETENTROPY
+	/* getentropy may exist but lack kernel support.  */
+	if (getentropy(&r, sizeof(r)) == 0)
+		return r;
+#endif
+
+#ifdef HAVE_GETRANDOM
+	/* Likewise getrandom.  */
+	if (getrandom(&r, sizeof(r), 0) == sizeof(r))
+		return r;
+#endif
+
+#ifdef HAVE_ARC4RANDOM_BUF
+	/* arc4random_buf can never fail.  */
+	arc4random_buf(&r, sizeof(r));
+	return r;
+#endif
+
+	/* Use /dev/urandom as a last resort.  */
+	fp = fopen("/dev/urandom", "r");
+	if (NULL == fp) {
+		goto fail;
+	}
+
+	if (fread(&r, sizeof(r), 1, fp) != 1) {
+		fclose(fp);
+		goto fail;
+	}
+
+	fclose(fp);
+	return r;
+
+fail:
+	fprintf(log_get_logfd(), _("Unable to obtain random bytes.\n"));
+	exit(1);
+}
+
+
+/*
+ * Return a uniformly-distributed random number smaller than upper_bound:
+ * [0, upper_bound)
+ *
+ * Similar to arc4random_uniform(3), but returns a u_long.
+ */
+unsigned long
+shadow_random_uniform(unsigned long upper_bound)
```

Comment:
The algo I implemented in Linux might be a bit more efficient if the cost of multiplication is low compared to the cost of drawing a new random sample, which is usually the case: 

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=e9a688bcb19348862afe30d7c85bc37c4c293471
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=7f576b2593a978451416424e75f69ad1e3ae4efe

Reply:
Hi Jason,

Hmmm, those Linux commits do too much magic to my taste.  Even with the comment and the commit message, I can't still understand why the multiplication works.

So, for an initial implementation, I'd keep the simple truncation with `&=`, even if it's slower.

After that, well, for shadow, I don't think the performance is worth making the code cryptic.  Moreover, since we use 64-bit numbers in this code, we would need 128-bit multiplication, which is not something supported everywhere.

However, it might be an interesting thing to do in glibc.  We could discuss improving the glibc implementation of arc4random_uniform(3) if you think it is an improvement over the current glibc code (please check).  If we do that, I suggest that you prepare a very detailed description of why the multiplication works, because at least in my case (and I suspect for many other programmers too), I already forgot the details of how integer multiplication works bitwise from when I studied that at school, and it's something I never use, so it's hard to understand why it serves the purpose here.

But, don't hurry; after Theo's email, I'm starting to strongly believe that the arc4random_uniform(3) interface is really bad, and you were right when you suggested a few months ago that we should reconsider if we want to add it, or maybe design a better API.  I will propose deprecating arc4random*(3) functions in glibc, and come up with a better set of interfaces, maybe implemented entirely in the kernel.

I'll also send some draft patches to the kernel, to start discussion about new functions that improve over the current get_random*() that you showed, and from that discussion we can maybe come up with new syscalls.

Cheers,

Alex

Reply:
Lemire has proofs: https://arxiv.org/abs/1805.10941

Your previous comment had a mention of expanding the kernel API. Not sure what you're proposing, but I've got a patchset under review that will make arc4random() and company both safe and fast.

Reply:
> Lemire has proofs: https://arxiv.org/abs/1805.10941

Thanks!

> Your previous comment had a mention of expanding the kernel API. Not sure what you're proposing,

I'll restore the message here, since it seems that arc4random*(3) can't be fixed because the bugs have percolated to users that now depend on the design issues.  That's fair, though; nothing surprising in C's history.

The removed part of my message was:

"
But, don't hurry; after Theo's email, I'm starting to strongly believe that the arc4random_uniform(3) interface is really bad, and you were right when you suggested a few months ago that we should reconsider if we want to add it, or maybe design a better API. I will propose deprecating arc4random*(3) functions in glibc, and come up with a better set of interfaces, maybe implemented entirely in the kernel.

I'll also send some draft patches to the kernel, to start discussion about new functions that improve over the current get_random*() that you showed, and from that discussion we can maybe come up with new syscalls.
"

> but I've got a patchset under review that will make arc4random() and company both safe and fast.

Please add me to the loop; I'm curious.  What's the link?

Reply:
I finally understood how the optimization works.  Here's some reduction to 3 bits, which (at least to me) clarifies how the rejection works.

![image](https://user-images.githubusercontent.com/21089465/212973388-27357904-5f1c-42a8-abb6-c5644df23dc0.png)

Since it is on average much faster than my approach, I'll apply it as an optimization, in a following patch.

Reply:
This is a graph of the probability of looping, for the algorithm with multiplication:

![image](https://user-images.githubusercontent.com/21089465/212994194-4e54c7fb-761c-4188-8c72-fce5e2417b61.png)

From those saw lines, only the last one goes up to 0.5.  As you go closer to 0, they are smaller.  This is compared to my implementation with masks, where _all_ the saw lines go up to 0.5.

Reply:
@ikerexxe , @hallyn 

This is a considerable performance improvement for low values of N.  The only downside is that it requires using a type wider than long, which would need to be `__int128`.  It's a GNU extension, so I don't know how much it's viable.  The current approach of masking is not so bad, it will normally just call a handful of times getrandom(2), but it could be much better.

Reply:
Here goes an (incomplete) 4-bit visual explanation of how it works, to compare it to the 3-bit one from above.  We can see that the rejection is smaller for low values of N compared to the same values of N when the original random number generator has less bits.

![image](https://user-images.githubusercontent.com/21089465/213282788-a78e3879-f933-4df2-a992-aa1508ea3af4.png)